lib/brakeman/differ.rb

```diff
@@ -26,26 +26,54 @@ def diff
   # only by line number.
   # Horrible O(n^2) performance.  Keep n small :-/
   def second_pass(warnings)
-    # keep track of the number of elements deleted because the index numbers
-    # won't update as the list is modified
-    elements_deleted_offset = 0
+    new_fingerprints = Set.new(warnings[:new].map(&method(:fingerprint)))
+    fixed_fingerprints = Set.new(warnings[:fixed].map(&method(:fingerprint)))
 
-    # dup this list since we will be deleting from it and the iterator gets confused.
-    # use _with_index for fast deletion as opposed to .reject!{|obj| obj == *_warning}
-    warnings[:new].dup.each_with_index do |new_warning, new_warning_id|
+    # Remove warnings which fingerprints are both in :new and :fixed
+    new_fingerprints.intersection(fixed_fingerprints).each do |hash|
```

Comment:
Would it be better to invert these loops?

I think the `delete_if` loops should be on the outside and they can check if the fingerprint is in the set.

Something like:

```ruby
shared_fingerprints = #...intersection
warnings[:new].delete_if do |warning
  shared_fingerprints.include? fingerprint(warning)
end
```
That way each array is only modified once.

You could also exit early if the intersection is empty.

Reply:
Good catch, thanks!  😅